### PR TITLE
Feature/test cases customer

### PR DESCRIPTION
**Pull request by sathish199308**
Test Case for Customers and Delete operation in API and MVC for Customer

### CODE REVIEW

File: BookingSystem/APIClient/Repository/APIClientCustomerRepository.cs [hunk 1]
```diff
@@ -63,25 +62,47 @@ public async Task<List<CustomerVM>> GetActiveCustomerAsync()
...
                 {
-                 
-                        var data = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
-                        var responseObject = JsonConvert.DeserializeObject<object>(data);
-                        var bookingData = JsonConvert.DeserializeObject<ResponseModel>(responseObject.ToString());
-                        var result = JsonConvert.DeserializeObject<CustomerVM>(bookingData.response.ToString());
-                        return result;
+                    var data = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                    var responseObject = JsonConvert.DeserializeObject<object>(data);
+                    var bookingData = JsonConvert.DeserializeObject<ResponseModel>(responseObject.ToString());
```
Comment (Backiaraj07) [inline BookingSystem/APIClient/Repository/APIClientCustomerRepository.cs:74]: Give relevant name to variables

File: BookingSystem/APIClient/Repository/APIClientCustomerRepository.cs [hunk 2]
```diff
@@ -63,25 +62,47 @@ public async Task<List<CustomerVM>> GetActiveCustomerAsync()
...
+            {
+                if (response.StatusCode == HttpStatusCode.OK)
+                {
+                    var data = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                    var responseObject = JsonConvert.DeserializeObject<object>(data);
+                    var bookingData = JsonConvert.DeserializeObject<ResponseModel>(responseObject.ToString());
+                    if(bookingData != null)
+                    {
+                        return true;
+                    }
```
Comment (Backiaraj07) [inline BookingSystem/APIClient/Repository/APIClientCustomerRepository.cs:101]: Give relevant name for variables

File: BookingSystem/Controllers/CustomerController.cs [hunk 3]
```diff
@@ -19,26 +19,29 @@ public class CustomerController : Controller
...
+        private const string saveCustomer = "SaveCustomer";
+        private const string getCustomerById = "GetCustomerById";
 
         public CustomerController(IAPIClientCustomerRepository clientCustomerRepository, IMapper mapper, ILoggerFactory logFactory)
         {
             _clientCustomerRepository = clientCustomerRepository;
             _mapper = mapper;
             _logger = logFactory.CreateLogger<CustomerController>();
         }
+
```
Comment (Backiaraj07) [inline BookingSystem/Controllers/CustomerController.cs:32]: Make logger and mapper first, then add repository

File: BookingSystem/Controllers/CustomerController.cs [hunk 4]
```diff
@@ -52,35 +55,58 @@ public async Task<IActionResult> Index()
...
+                _logger.LogInformation(saveCustomer + complete);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex.Message);
+            }
+
+            _logger.LogInformation(nameof(CustomerController) + "." + nameof(Customer) + complete);
+
+            return RedirectToAction("Index");
```
Comment (Backiaraj07) [inline BookingSystem/Controllers/CustomerController.cs:76]: Use Name of function to redirect to action method.

File: BookingSystem/Controllers/CustomerController.cs [hunk 5]
```diff
@@ -52,35 +55,58 @@ public async Task<IActionResult> Index()
...
         }
+
+        [HttpGet]
+        public IActionResult DeleteCustomer(int? CustomerID)
+        {
+            if (CustomerID > 0)
+            {
+                 _clientCustomerRepository.DeleteCustomer(CustomerID ?? 0);
+            }
+            return RedirectToAction("Index");
```
Comment (Backiaraj07) [inline BookingSystem/Controllers/CustomerController.cs:109]: Use Nameof function to redirect to action method.

File: BookingSystemAPI/Controllers/CustomerController.cs [hunk 6]
```diff
@@ -22,6 +22,8 @@ public class CustomerController : ControllerBase
         private const string nullCustomer = "NullCustomer";
         private const string saveCustomer = "SaveCustomerDetails";
         private const string getcustomerbyId = "GetCustomerDetailsById";
+        private const string deletecustomerbyId = "DeleteCustomerById";
+
         public CustomerController(ICustomerService customerService, ILoggerFactory logFactory)
         {
             _customerService = customerService;
```
Comment (Backiaraj07) [inline BookingSystemAPI/Controllers/CustomerController.cs:29]: Declare Iloggerfactory first, then repository

File: BookingSystem.DAL/Repository/CustomerRepository.cs [hunk 7]
```diff
@@ -103,6 +102,17 @@ public async Task<CustomerVM> GetActiveCustomerByCustomerIdAsync(long CustomerID
             return customer;
         }
 
-       
+        public bool DeleteCustomer(int CustomerID)
```
Comment (manic1990) [inline BookingSystem.DAL/Repository/CustomerRepository.cs:105]: Use async method.

File: BookingSystemAPI/Controllers/CustomerController.cs [hunk 8]
```diff
@@ -108,21 +108,47 @@ public async Task<ActionResult> GetCustomerById(long? customerId)
...
-
+        }
 
+        [HttpDelete]
 
-        }
+        [Route("deletecustomerbyId")]
+        public async Task<IActionResult> DeleteCustomer(int CustomerID)
+        {
+            _logger.LogInformation(nameof(CustomerController) + nameof(GetCustomerById), started);
```
Comment (Backiaraj07) [inline BookingSystemAPI/Controllers/CustomerController.cs:125]: give the proper method name while log

File: BookingSystemAPI/Controllers/CustomerController.cs [hunk 9]
```diff
@@ -108,21 +108,47 @@ public async Task<ActionResult> GetCustomerById(long? customerId)
...
+                    response.statuscode = HttpStatusCode.InternalServerError;
+                    response.errordescription = ex.Message.ToString();
+                }
 
+                _logger.LogInformation(nameof(CustomerController) + nameof(getcustomerbyId), completed);
+            }
+            else
+            {
+                _logger.LogInformation(nameof(CustomerController) + nameof(getcustomerbyId), nullCustomer);
+            }
```
Comment (Backiaraj07) [inline BookingSystemAPI/Controllers/CustomerController.cs:150]: Log the method name correctly

**Review by Backiaraj07 — COMMENTED** [review]
Fix the comments

File: BookingSystem.Tests/BAL.Tests/CustomerServiceTest/SaveCustomerTest.cs [hunk 10]
```diff
@@ -27,14 +23,13 @@ public SaveCustomerTest()
         }
 
         [Fact]
-
         public async Task SaveBooking_ReturnOKResult_WithBookingId()
```
Comment (manic1990) [inline BookingSystem.Tests/BAL.Tests/CustomerServiceTest/SaveCustomerTest.cs:26]: Use proper method name.

File: BookingSystem.Tests/BAL.Tests/CustomerServiceTest/UpdateCustomerTest.cs [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+        public UpdateCustomerTest()
+        {
+            _customerRepository = new Mock<ICustomerRepository>();
+            _addressRepository = new Mock<IAddressRepository>();
+            fixture = new Fixture();
+        }
+
+        [Fact]
+        public async Task SaveBooking_ReturnOKResult_WithBookingId()
```
Comment (manic1990) [inline BookingSystem.Tests/BAL.Tests/CustomerServiceTest/UpdateCustomerTest.cs:26]: User proper method name.

File: BookingSystem/APIClient/Repository/APIClientCustomerRepository.cs [hunk 12]
```diff
@@ -63,25 +62,47 @@ public async Task<List<CustomerVM>> GetActiveCustomerAsync()
...
-                 
-                        var data = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
-                        var responseObject = JsonConvert.DeserializeObject<object>(data);
-                        var bookingData = JsonConvert.DeserializeObject<ResponseModel>(responseObject.ToString());
-                        var result = JsonConvert.DeserializeObject<CustomerVM>(bookingData.response.ToString());
-                        return result;
+                    var data = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                    var responseObject = JsonConvert.DeserializeObject<object>(data);
+                    var bookingData = JsonConvert.DeserializeObject<ResponseModel>(responseObject.ToString());
+                    var result = JsonConvert.DeserializeObject<CustomerVM>(bookingData.response.ToString());
```
Comment (manic1990) [inline BookingSystem/APIClient/Repository/APIClientCustomerRepository.cs:75]: Before DeserializeObject check null for bookingData.response.

File: BookingSystem/Controllers/CustomerController.cs [hunk 13]
```diff
@@ -19,26 +19,29 @@ public class CustomerController : Controller
...
+                _logger.LogInformation(getCustomer + start);
 
                 var result = await _clientCustomerRepository.GetActiveCustomerAsync().ConfigureAwait(false);
                 var viewCustomer = _mapper.Map<List<ViewCustomerData>>(result);
 
-                _logger.LogInformation(GetCustomer + complete);
+                _logger.LogInformation(getCustomer + complete);
 
                 return View(viewCustomer);
             }
```
Comment (manic1990) [inline BookingSystem/Controllers/CustomerController.cs:47]: use catch block

File: BookingSystemAPI/Controllers/CustomerController.cs [hunk 14]
```diff
@@ -108,21 +108,47 @@ public async Task<ActionResult> GetCustomerById(long? customerId)
                     response.errordescription = ex.Message.ToString();
                 }
```
Comment (manic1990) [inline BookingSystemAPI/Controllers/CustomerController.cs:109]: use catch block.

File: BookingSystem.DAL/Repository/CustomerRepository.cs (hunk 7 above)
Comment (sathish199308) [reply BookingSystem.DAL/Repository/CustomerRepository.cs:105]: Change the bool method to Asynchronous is done in all the layers.

File: BookingSystem/APIClient/Repository/APIClientCustomerRepository.cs (hunk 1 above)
Comment (sathish199308) [reply BookingSystem/APIClient/Repository/APIClientCustomerRepository.cs:74]: Naming variable change is done

File: BookingSystem/APIClient/Repository/APIClientCustomerRepository.cs (hunk 2 above)
Comment (sathish199308) [reply BookingSystem/APIClient/Repository/APIClientCustomerRepository.cs:101]: Naming Variable change is done

File: BookingSystem/Controllers/CustomerController.cs (hunk 3 above)
Comment (sathish199308) [reply BookingSystem/Controllers/CustomerController.cs:32]: Changes made an Input Parameters

File: BookingSystem/Controllers/CustomerController.cs (hunk 4 above)
Comment (sathish199308) [reply BookingSystem/Controllers/CustomerController.cs:76]: Nameof Method is implemented

File: BookingSystem/Controllers/CustomerController.cs (hunk 5 above)
Comment (sathish199308) [reply BookingSystem/Controllers/CustomerController.cs:109]: Nameof Method is implement

File: BookingSystemAPI/Controllers/CustomerController.cs (hunk 6 above)
Comment (sathish199308) [reply BookingSystemAPI/Controllers/CustomerController.cs:29]: Implemnted Successfully

File: BookingSystemAPI/Controllers/CustomerController.cs (hunk 8 above)
Comment (sathish199308) [reply BookingSystemAPI/Controllers/CustomerController.cs:125]: Proper name is given

File: BookingSystemAPI/Controllers/CustomerController.cs (hunk 9 above)
Comment (sathish199308) [reply BookingSystemAPI/Controllers/CustomerController.cs:150]: Method name change is done

File: BookingSystem.Tests/BAL.Tests/CustomerServiceTest/SaveCustomerTest.cs (hunk 10 above)
Comment (sathish199308) [reply BookingSystem.Tests/BAL.Tests/CustomerServiceTest/SaveCustomerTest.cs:26]: Proper method name is given

File: BookingSystem.Tests/BAL.Tests/CustomerServiceTest/UpdateCustomerTest.cs (hunk 11 above)
Comment (sathish199308) [reply BookingSystem.Tests/BAL.Tests/CustomerServiceTest/UpdateCustomerTest.cs:26]: Proper method name is given

File: BookingSystem/APIClient/Repository/APIClientCustomerRepository.cs (hunk 12 above)
Comment (sathish199308) [reply BookingSystem/APIClient/Repository/APIClientCustomerRepository.cs:75]: Null Data checking is implement successfully before Deserialize

File: BookingSystem/Controllers/CustomerController.cs (hunk 13 above)
Comment (sathish199308) [reply BookingSystem/Controllers/CustomerController.cs:47]: Catch block is implemented.

File: BookingSystemAPI/Controllers/CustomerController.cs (hunk 14 above)
Comment (sathish199308) [reply BookingSystemAPI/Controllers/CustomerController.cs:109]: Catch block is implemented

**Review by Backiaraj07 — APPROVED** [review]
Verified

**Pull request merged**